include/openssl/core_dispatch.h

```diff
@@ -684,6 +699,10 @@ OSSL_CORE_MAKE_FUNC(const OSSL_PARAM *, keymgmt_import_types_ex,
 OSSL_CORE_MAKE_FUNC(const OSSL_PARAM *, keymgmt_export_types_ex,
                     (void *provctx, int selection))
 
+# define OSSL_FUNC_KEYMGMT_RESERVE                    47
+OSSL_CORE_MAKE_FUNC(void *, keymgmt_reserve,
+                    (void *keydata, uint64_t count))
+
```

Comment:
I noticed this, and that peaked my curiosity...  and to be frank, a bit of scepticism, but that is no longer the case.

Having seen what it does, I get it...  but oh boy, it's so irritatingly close to a `dup`!  Has me wishing we had thought of the possibility of passing parameters to the `dup` function.  Ah well, we'll have to have a special case.

That being said, do we really want it to be *this* special?  After all, it's very specifically tailored for HSS keys, isn't it?  Would it be very hard to actually make it a parametrized `dup`, just so that it becomes a bit more agnostic to the specific algos?  Something like this, perhaps?

```suggestion
# define OSSL_FUNC_KEYMGMT_RESERVE                    47
OSSL_CORE_MAKE_FUNC(void *, keymgmt_dup_with_params,
                    (const void *keydata, OSSL_PARAM params[]))

```

Reply:
If anything, it should be a dup_ex function also with added `selection` parameter. This could be used to implement "duplicate me a public key from a keypair".

Reply:
I might not have given this a good name yet, but this isnt really just a dup.
Its more like splitting the key into 2 parts..
The key consists of a set of keypairs that can only be used once for signing. So the sign is problematic if a key is used again.
This creates all sorts of problems, so this 'reserve' attempts to grab just a shard of the keypairs. If there was some failure during this time, then this shard would  just be thrown away.
I havent got far with XMSS yet, but I believe it will need the same mechanism.

Reply:
This function is also very useful for testing, since I can skip forward through a larger tree to test things.

Reply:
I'm thinking of it from a general API perspective, and it sure looks like duplication *even if there are internal/side effects that are specific to the algorithm*.

The general view is that you get a new key back from passing an existing key.

Reply:
What about from a user point of view? Using EVP_PKEY_dup_ex() to shard a key seems strange to me.

Reply:
Note that doing just a dup of this type of key is a really dangerous operation, so without any parameters it would either need to fail or grab everything..

Reply:
> What about from a user point of view? Using EVP_PKEY_dup_ex() to shard a key seems strange to me.

It's possible that we can have a better general name, I'm not opposed to that...  Either way, whatever the effects are on the new key as well as on the original key needs to be documented in the algorithm specific documentation.

For a bit, my thoughts went to cell division (biology), but I'm not sure I'd like that as a name for this function.

Generality is *hard*

Reply:
> Note that doing just a dup of this type of key is a really dangerous operation, so without any parameters it would either need to fail or grab everything..

So if you have an internal default for the count set to zero, and that it remains zero after having processed parameters, you return an error, wouldn't that be the right thing to do?

Reply:
I had to take a closer look at the RFC to understand this thing, and found my way to Cisco's reference implementation, where I found a "reserve" functionality that is completely different...  is it the same thing that you're trying to achieve, but you do it in a different manner?

Ref:

https://github.com/cisco/hash-sigs/blob/2020bc16f3ab12d8c1da1e92749dc2874a5e860e/hss_reserve.c#L101-L113

Reply:
I made no attempt to copy the cisco code. i.e. It was done from scratch.
I also looked at the Bouncy castle implementation of HSS.
I am also not claiming to have solved the non volatile requirements that hardware signing would require.

Reply:
I understand, but none of that is what I asked about.  I'm trying to understand where this "reserve" thing comes from, conceptually.

Reply:
So it is based on what BC did (they have a private key method called extractKeyShard), which is based off the concept in the ref implementation.
I tried explaining it above.. i.e.
 
 A HSS/LSS tree can only be used a fixed number of times for signing.
The hierarchy of LMS trees has key pairs in the leaf nodes of the trees. These can be quite large.
The private key has state in it which contains the next leaf index, that is incremented during each sign operation.
Dealing with this 'state', which would need to be persisted is not trivial  (rebooting for example). See SP800-208 Section 8.1.

So this code takes the approach of giving you an option to extract a smaller portion of the private keys, and then just use this smaller portion. In the event of an error/failure you could then throw away this smaller portion of keys, and resume with the remaining larger portion of the keys. You would then grab another shard off the remaining keys.

Reply:
Basically, the algorithm needs to reserve space in all the parent trees before it can commit to signing anything.
Any reuse anywhere completely breaks things and this is the only way that guarantees no reuse (under sufficiently weak assumptions).

So essentially the signing must reserve & commit everything higher than the bottom layer before finishing any signing operation.  This is best done well before the final signing to avoid later contention.  In the event of a failure, this leaves gaps which have to be lived with -- they should be harmless.  That's still far better than a reuse.  Re-read that thinking "highly multi-threaded" and then re-read it thinking many separate process or machines.  It's somewhat quixotic, there is no win here.

This is the fundamental reason why @slontis didn't want to support signing.  It is really really difficult to absolutely guarantee single use.  Almost impossible in fact.  I believe we should abrogate that responsibility as much as possible.  However, the verify only option was trashed during an OTC meeting some while ago.  Which leaves us with a flawed & possibly unsafe signing operation that's only relevant for hardware builds.  Software implementations must never expose the signing operation.  Moreover, NIST has banned software modules from supporting signing.

Reply:
What about completely disabling signing support on builds with enabled threads?

There is one major reason why naming the operation dup_ex would be bad as the source key is being modified! by this reserve operation. Any real duplicate operation (even if there are some options/params) would not modify the source key.

Reply:
Thank you for the thorough explanations.  That was quite enlightening, and I can see how it needs to be conceptually separate from `dup`.

I do still worry about the count argument.  Yes, that works for LMS / HSS specifically, but will that work for the next algorithm that wants to use a similar concept, but with different parameters?  It becomes untenable if keymgmt starts piling up algo specific functions, all conceptually doing the same thing, just driven by different parameters...

Reply:
In the semantic sense the reserve operation is actually very similar to generate operation when generating a key from parameters as the primary private key cannot be used for any real operation. However it again modifies the primary key where domain parameters for other algorithms are not modified when keys are generated from them.

Reply:
@t8m this isnt just a threading issue.. Read SP800-208 Section 9.1.

@levitte  I can make it take a param if that helps..

Reply:
> @levitte I can make it take a param if that helps..

I think that would be a good enough level of future proofing.

Reply:
@t8m you can use either part of the split key (you dont have to split at all if you are not worried about state changes).
The keygen doesnt have to actually do much apart from create a single active tree for each level. Apon any sign operation the key state (q index) is updated. Once the key updates, an active tree can become full, so it would then create a new active tree. For a new active tree, It then goes up to the parent tree and increments that one also. (At some point that parent tree will become full, so it creates a new active tree and propagates upwards). The top level tree is the only tree that is fully created (and if you reach the end of that one you have no more keys).

Reply:
Another thought on the subject....  why does this function need to be exposed?  Would it be insane to just have the signature init reserve a count of 1, completely internally in the provider?

Reply:
Hmm, actually why not call the operation `split`? And I agree with @levitte that instead of int argument OSSL_PARAM array should be used.

Reply:
Heh...  "split" has my brain go "Oh!  Shamir's secret sharing!"

Not that I mind, really, but it does tell that this concept will have to have good conceptual documentation.